File: CodingInterviewsPythonSolutions/138. Copy List with Random Pointer.py
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None

        # Create a mapping of original nodes to their corresponding copies
        node_map = {}
        cur = head
        while cur:
            node_map[cur] = Node(cur.val)
            cur = cur.next

        # Assign next and random pointers for the copy nodes based on the mapping
        cur = head
        while cur:
            copy_node = node_map[cur]
            copy_node.next = node_map.get(cur.next)
            copy_node.random = node_map.get(cur.random)
            cur = cur.next

        return node_map[head]

File: CodingInterviewsJavaSolutions/138. Copy List with Random Pointer.java
class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }

        Node result = null;
        Node new_result = null;
        HashMap<Node, Node> my_map = new HashMap<>();

        // First pass: Create new nodes and map original nodes to new nodes
        for (Node temp_head = head; temp_head != null; temp_head = temp_head.next) {
            Node new_node = new Node(temp_head.val);
            new_node.random = temp_head.random;
            my_map.put(temp_head, new_node);

            if (result == null) {
                result = new_node;
            } else {
                new_result.next = new_node;
            }

            new_result = new_node;
        }

        // Second pass: Adjust random pointers using the map
        for (Node temp_head = result; temp_head != null; temp_head = temp_head.next) {
            temp_head.random = my_map.get(temp_head.random);
        }

        return result;
    }
}
